File: 122-best-time-to-buy-and-sell-stock-ii/122-best-time-to-buy-and-sell-stock-ii.java
class Solution {
    public int maxProfit(int[] prices) {
    int ahead[]=new int[2];
    int cur[] =new int[2];
        
    int n = prices.length;
    ahead[0] = ahead[1] = 0;
        
    int profit = 0;
    
    for(int idx = n-1;idx >=0;idx--){
        for(int buy = 0; buy<=1; buy++){
            if(buy == 0){
                profit = Math.max(-prices[idx]+ahead[1],ahead[0]);
            }
            
            if(buy == 1)
            {
                profit = Math.max(prices[idx]+ahead[0],ahead[1]);
            }

                cur[buy]  = profit;
        } 
        
        ahead = (int[])(cur.clone());
    }
        
        return cur[0];
    }
}

File: 122-best-time-to-buy-and-sell-stock-ii/122-best-time-to-buy-and-sell-stock-ii.js
/**
 * @param {number[]} prices
 * @return {number}
 */

function f(idx,buy,prices,dp){
    if(idx == prices.length) return 0;
    
    if(dp[buy][idx] !=-1) return dp[buy][idx];
    
    let profit = 0;
    if(buy == 1){
        profit = Math.max(-prices[idx]+f(idx+1,0,prices,dp),f(idx+1,1,prices,dp));
    }else{
        profit = Math.max(prices[idx]+f(idx+1,1,prices,dp),f(idx+1,0,prices,dp));
    }
    
    return dp[buy][idx] = profit;
}

var maxProfit = function(prices) {
    const dp = new Array(2).fill(0).map(p => new Array(prices.length).fill(-1));
    // console.log(dp);
    
    return f(0,1,prices,dp);
    
};